### DISCUSSION TITLE
Python asyncio 簡單教學

### DISCUSSION THREAD

**Discussion by ATREE01** [1101-GS4719-C/Python-Community-GS4719-C#1702] [General]
#    Python asyncio

#    前言
這篇文章將要教學如何使用Python的asyncio套件，這是一個很好用但也有點難的東西，可以讓你的程式同步執行，提升程式效率。

#    甚麼是asyncio
我們先來問問ChatGPT
**Use Case:** 
Well-suited for I/O-bound tasks where tasks spend a significant amount of time waiting for external resources (e.g., network requests, file I/O).
**Advantages:**
Single-threaded event loop allows for efficient context switching.
Good for handling a large number of simultaneous connections.
It can be more scalable than traditional synchronous programming for I/O-bound tasks.
**Disadvantages:**
Not ideal for CPU-bound tasks.
Requires understanding and adherence to the asynchronous programming model.

簡單來說就是適合用於程式可能會花比較多的時間在等待外部資源的處理的情況下，可以透過這個套件的設計讓你的程式執行速度變快。

#    本文

假設我們今天有三個任務要執行，一個需要花費1秒，一個兩秒，一個三秒。程式碼如下

```python
import time

def task1():
    print("Starting task 1")
    time.sleep(1)
    print("task 1 finished")
    
def task2():
    print("Starting task 2")
    time.sleep(2)
    print("task 2 finished")
    
def task3():
    print("Starting task 3")
    time.sleep(3)
    print("task 3 finished")

if __name__ == "__main__":
    startTime = time.time()
    task1()
    task2()
    task3()
    print("Total time: ",time.time() - startTime)
```
執行結果:

![image](https://hackmd.io/_uploads/S17ojCW4p.png)

我們總共需要花費6秒的時間，但是假如這三個任務個字是彼此之間是互相獨立不會互相影響的，我們就可以透過asyncio來讓這些任務能夠同步執行。

在開始之前我們需要先介紹兩個東西:

`async`:這個字是非同步的縮寫，我們會說透過這個方法定義的函式叫做`coroutine`(協程)能夠被同步執行，使用的方法也很簡單，在宣告函式時加入即可。
`await`:則是宣告在可以被等待的物件前，這個字的意思式是宣告這行程式的內容是要被等待的，在等待的期間可以去執行其他的coroutine。至於有那些是可以被等待的想要深入研究的話可以參考[Awaitables](https://docs.python.org/3/library/asyncio-task.html#awaitables)像是async函式就是一種可以被等待的函式。

接下來我們把原本的程式碼修改成以下

```python
import time
import asyncio

async def task1():
    print("Starting task 1")
    await asyncio.sleep(1)
    print("task 1 finished")
    
async def task2():
    print("Starting task 2")
    await asyncio.sleep(2)
    print("task 2 finished")
    
async def task3():
    print("Starting task 3")
    await asyncio.sleep(3)
    print("task 3 finished")

```

在修改後我們呼叫函式的方式也要修改，才能夠發揮它的功用。

```python
if __name__ == "__main__":
    async def main():
        startTime = time.time()
        tasks = [asyncio.create_task(task1()), asyncio.create_task(task2()), asyncio.create_task(task3())]
        for task in tasks:
            await task        
        print("Total time: ",time.time() - startTime)
    asyncio.run(main())
```

`asyncio.run`:This function runs the passed coroutine, taking care of managing the asyncio event loop, finalizing asynchronous generators, and closing the executor.
簡單來說他會幫你管理並執行你給他的協程，不用自己去做一下比較低等級的互動。

`create_task`:Wrap the coro coroutine into a Task and schedule its execution
`Task`:Tasks are used to schedule coroutines concurrently.

簡單理解上面的意思就式create_task會把coroutine包裝成一個task，而task就是可以同步執行Coroutine的關鍵。

如果我們執行上面的程式碼會看到以下結果:

![image](https://hackmd.io/_uploads/BkZQM1MV6.png)
會發現原本需要6秒的執行時間現在變成3秒這是因為整個流程變成以下
執行第一個task遇到await開始等待1秒，在這個時候他會先跳去執行第二個task，同樣遇到await所以跳去執行第三個task。

這三件事是同步發生的，所以最終只要等待3秒，而原本程式則是執行完第一個之後才會去執行第二個因此需要6秒

如果把程式碼更改成以下會更方便理解:
```python=
async def task1():
    print("Starting task 1")
    await asyncio.sleep(3)
    print("task 1 finished")
    
async def task2():
    print("Starting task 2")
    await asyncio.sleep(2)
    print("task 2 finished")
    
async def task3():
    print("Starting task 3")
    await asyncio.sleep(1)
    print("task 3 finished")
```

執行結果
![image](https://hackmd.io/_uploads/Hkg4EmJzE6.png)

可以看到雖然是第一個task先執行，但是因為他要等待三秒的關係，所以是第三個先執行完。

#    結語
asyncio本身是一個好用的東西，而且在其他的語言中也會有這種await跟async的概念可以使用，雖然學習難度有點高，但是學會之後絕對會很方便，還可以讓你的程式更有效率。我這篇文章只是很簡單的講解，有很多不足之處，如果想要進一步的學習可以多看看其他人的文章，或是參考官方的[document](https://docs.python.org/zh-tw/3/library/asyncio.html)。

**Comment by 1chooo** [comment]
感謝你的分享，`asyncio` 確實是個非常好用的套件，尤其是我在網頁開發相關的專案之中非常常用到，如此可以讓執行不用卡在一起，而是可以同時處理不同的工作！

**Comment by lazywulf** [comment]
超棒的！我記得第一次看這鬼完全不知道這啥。對非同步程式除錯也是非常頭痛。
這東西當初花我好多時間，就只為了把爬蟲速度提起來。

**Comment by BobomanOUOb** [comment]
我自己在使用python的multiprocess時，我自己是使用另一個package 叫做 multiprocessing ，然後再從這個package import pool 把不同的程式片段分到不同的pool ，在交由CPU的不同核心來去執行，詳細的介紹可以看看這篇文章 https://docs.python.org/zh-tw/3/library/multiprocessing.html
或是可以用一些神奇的方法來提高python code 的運行優先級，詳細的可以參考此網站: https://infoacetech.net/windows/cpu%E5%84%AA%E5%85%88%E9%A0%86%E5%BA%8F/#google_vignette 
基本上要進到控制台去更改設定，但注意CPU的其中一核為處理開機程序的。
